wallet: Refactor and document CoinControl

aureleoules opened the pull request:
- Moves CoinControl function definitions from `coincontrol.h` to `coincontrol.cpp`
- Adds more documentation
- Renames class member for an improved comprehension
- Use `std::optional` for `GetExternalOutput`

aureleoules committed:
wallet: Use std::optional for GetExternalOutput and fixups

aureleoules committed:
wallet: Move CoinCointrol definitions to .cpp

Move definitions to coincontrol.cpp and add documentation.

yancyribbens commented:
These changes improve the code quality and implement the recommended practices described in the [developer notes - source code organization](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md?plain=1#L1051).

Concept ACK

w0xlt reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/26066/commits/ae67293025392e8fb5054b07a5173209c793ed6c

furszy reviewed (commented):
In dba7f9b5:
Why only change `IsSelected` and `HasSelected` and not the setters? Should be consistent.

Saying that, even if we are consistent on the renaming, I'm not so sure about it. Do you think that `PreSelect(outpoint)`/`IsPreSelected(outpoint)` is more understandable than `Select(outpoint)`/`IsSelected(output)`?

At least in my view, it doesn't look more readable than before.
Would instead aim for a good docstring and done.

aureleoules committed:
scripted-diff: Rename setSelected->m_selected_inputs

-BEGIN VERIFY SCRIPT-
sed -i 's/setSelected/m_selected_inputs/g' src/wallet/coincontrol.h src/wallet/coincontrol.cpp
-END VERIFY SCRIPT-

aureleoules commented:
@furszy I've slightly improved the docstring of `Select` and `SelectExternal` and rolled back to `IsSelected` and `HasSelected`.
I kept the old `setSelected` as `m_preset_inputs` though.

Is it better?

furszy reviewed (commented):
Yeah @aureleoules better 👍🏼 .

aureleoules commented:
All cleaned-up @furszy!

furszy commented on src/wallet/coincontrol.h:
```diff
@@ -63,73 +63,56 @@ class CCoinControl
+     */
+    std::optional<CTxOut> GetExternalOutput(const COutPoint& outpoint) const;
+    /**
+     * Pre-selects the given output.
+     * The output will be included in the transaction even if it's not the most optimal choice.
```
nit: would remove the "pre-selects" term and go directly with something like "store the given outpoint."

I don't think that the "pre" term here provides information to readers. (the second paragraph is great to express what the process will do with the selected outpoint)

(same for the others methods as well).

aureleoules replied:
How about this?
```suggestion
     * Lock-in the selected output for spending.
     * The output will be included in the transaction even if it's not the most optimal choice.
```

furszy replied:
yeah, sounds good 👌🏼 .

furszy reviewed (approved):
ACK 9d4cfd6c

satsie commented on src/wallet/coincontrol.h:
```diff
@@ -63,76 +63,59 @@ class CCoinControl
 private:
-    std::set<COutPoint> setSelected;
+    std::set<COutPoint> m_preset_inputs;
     std::map<COutPoint, CTxOut> m_external_txouts;
```
nit: you've done such a fabulous job documenting the header methods (no good deed goes unpunished!) that I think I understand an external input is one that does not belong to the wallet. 

If you find yourself in the code again, is there any chance to get a comment on this data structure describing what it is? Something like `//! Map of inputs that do not belong to the wallet` (if that is indeed the correct definition). Actually... just for my own learning, is that the correct definition?

aureleoules replied:
Done!
And yes you are correct :) The wallet would not know how to sign these inputs but they are in the utxo set.

satsie commented:
ACK 9d4cfd6c5a40cafaeed31d0102b45a7e44dc0076, reviewed diff and ran unit + functional tests

I'm very new to the project but this was a great PR to review, especially after the changes that came in from earlier code review. Thanks for the comments you added on all the function definitions. Can confirm that they are helpful to new eyes and as far as I can tell, the changes are a nice cleanup that make things easier to read, understand, and follow the recommended practices (as mentioned in an earlier comment).

aureleoules commented:
Thanks for the reviews, I addressed your comments @satsie @furszy.

furszy reviewed (approved):
ACK 287719a

aureleoules commented:
sorry I failed my rebase @furszy

furszy commented:
or well, not ACK :p.
You forgot to change the `m_preset_inputs` inside `CCoinControl`

satsie commented:
ACK 5db911a369df3977c79e6d8f06ccc38551db9efb

furszy reviewed (approved):
ACK 5db911a3

aureleoules commented:
I think this is ready for merging? @w0xlt if you could reACK that would be great :smile:.

aureleoules commented:
Rebased.

theStack reviewed (approved):
Code-review ACK 22801dff82d5d5329a75cb7fc14bd6e05dc67a5b

Nice cleanup and documentation improvement!
nit: I have one more suggestion for an improvement, the interface of `CCoinControl::ListSelected` would be more straight-forward if the result is returned directly instead of using an out-parameter, simplifying the caller-side:

<details>
  <summary>Diff</summary>

```diff
diff --git a/src/qt/coincontroldialog.cpp b/src/qt/coincontroldialog.cpp
index bd9a90a89..4b9ed9028 100644
--- a/src/qt/coincontroldialog.cpp
+++ b/src/qt/coincontroldialog.cpp
@@ -412,8 +412,7 @@ void CoinControlDialog::updateLabels(CCoinControl& m_coin_control, WalletModel *
     unsigned int nQuantity      = 0;
     bool fWitness               = false;

-    std::vector<COutPoint> vCoinControl;
-    m_coin_control.ListSelected(vCoinControl);
+    auto vCoinControl{m_coin_control.ListSelected()};

     size_t i = 0;
     for (const auto& out : model->wallet().getCoins(vCoinControl)) {
diff --git a/src/wallet/coincontrol.cpp b/src/wallet/coincontrol.cpp
index 9837bb442..009e95c2c 100644
--- a/src/wallet/coincontrol.cpp
+++ b/src/wallet/coincontrol.cpp
@@ -58,9 +58,9 @@ void CCoinControl::UnSelectAll()
     m_selected_inputs.clear();
 }

-void CCoinControl::ListSelected(std::vector<COutPoint>& vOutpoints) const
+std::vector<COutPoint> CCoinControl::ListSelected() const
 {
-    vOutpoints.assign(m_selected_inputs.begin(), m_selected_inputs.end());
+    return {m_selected_inputs.begin(), m_selected_inputs.end()};
 }

 void CCoinControl::SetInputWeight(const COutPoint& outpoint, int64_t weight)
diff --git a/src/wallet/coincontrol.h b/src/wallet/coincontrol.h
index 0ff2ced72..953e20ab4 100644
--- a/src/wallet/coincontrol.h
+++ b/src/wallet/coincontrol.h
@@ -100,7 +100,7 @@ public:
     /**
      * List the selected inputs.
      */
-    void ListSelected(std::vector<COutPoint>& vOutpoints) const;
+    std::vector<COutPoint> ListSelected() const;
     /**
      * Set an input's weight.
      */
diff --git a/src/wallet/spend.cpp b/src/wallet/spend.cpp
index 2da95f037..c5deb8d33 100644
--- a/src/wallet/spend.cpp
+++ b/src/wallet/spend.cpp
@@ -145,9 +145,7 @@ util::Result<PreSelectedInputs> FetchSelectedInputs(const CWallet& wallet, const
                                             const CoinSelectionParams& coin_selection_params) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet)
 {
     PreSelectedInputs result;
-    std::vector<COutPoint> vPresetInputs;
-    coin_control.ListSelected(vPresetInputs);
-    for (const COutPoint& outpoint : vPresetInputs) {
+    for (const COutPoint& outpoint : coin_control.ListSelected()) {
         int input_bytes = -1;
         CTxOut txout;
         if (auto ptr_wtx = wallet.GetWalletTx(outpoint.hash)) {
```
</details>

theStack committed:
refactor: Make ListSelected return vector

aureleoules commented:
Thanks for the suggestion @theStack. I pushed b76fbbe8bf9504d8c08cc2f1281782db5393c02d and gave you authorship.

theStack reviewed (approved):
ACK b76fbbe8bf9504d8c08cc2f1281782db5393c02d

furszy commented:
While I would like to merge this one (it's good as is), I think that would be better to check #25273 first. Mainly because this PR is about improving the `CoinControl` class documentation and #25273 modifies a big part of it (which, if we merge this one first, will make us re-do this doc improvements in another follow-up later).

furszy commented:
@MarcoFalke, guess that Drahtbot shouldn't have requested review to me.

MarcoFalke commented:
Yeah, if it was live 3 weeks ago, it would have requested a review from you before your comment. Which, I guess, would have made sense.

Also, it will only request reviews when there is an email notification sent out anyway, so the harm/risk should be minimal. But let me know if it should be tamed down or turned off.

achow101 commented:
Are you still working on this?

aureleoules commented:
Rebased

murchandamus reviewed (approved):
ACK daba95700b0b77a2e898299f218c47a69ed2c7d0

achow101 commented:
ACK daba95700b0b77a2e898299f218c47a69ed2c7d0
